src/tqec/compile/observables/abstract_observable.py

```diff
@@ -0,0 +1,215 @@
+"""Describe the location of the observable measurements in the block graph."""
+
+from dataclasses import dataclass
+
+from tqec.compile.specs.enums import JunctionArms
+from tqec.computation.correlation import CorrelationSurface
+from tqec.computation.cube import Cube, ZXCube
+from tqec.computation.pipe import Pipe
+from tqec.position import Direction3D, Position3D
+from tqec.computation.block_graph import BlockGraph
+from tqec.computation.zx_graph import ZXEdge, ZXKind
+
+
+@dataclass(frozen=True)
+class AbstractObservable:
+    """An abstract description of a logical observable in the
+    :py:class:`~tqec.computation.BlockGraph`.
+
+    A logical observable corresponds to all the
+    `OBSERVABLE_INCLUDE <https://github.com/quantumlib/Stim/blob/main/doc/gates.md#OBSERVABLE_INCLUDE>`_
+    instructions with the same observable index in the ``stim`` circuit
+    and is composed of a set of measurements. Abstract observable
+    specifies where are the measurements located in the block graph.
+
+    Attributes:
+        top_readout_cubes: A set of cubes of which a straight line of data
+            qubit readouts on the top face should be included in the observable.
+        top_readout_pipes: A set of pipes of which a single data qubit readout
+            on the top face should be included in the observable.
+        bottom_stabilizer_pipes: A set of pipes of which a region of stabilizer
+            measurements on the bottom face(in the cubes it connects) should be
```

Comment:
```suggestion
            measurements on the bottom face (in the cubes it connects) should be
```
and I do not understand the information added in parenthesis.

Reply:
Pipes don't have spacetime extension, so stabilizer measurements take place in the cubes instead. I'll rephrase it slightly to make it clearer.

Reply:
Got it thanks!